Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -421,33 +397,60 @@ package final actor SemanticIndexManager {
     let sourceFiles = await orLog("Getting source files in project") {
       Set(try await buildSystemManager.sourceFiles(includeNonBuildableFiles: false).keys)
     }
     guard let sourceFiles else {
       return []
     }
-    let filesToReIndex = await files.asyncCompactMap { (uri) -> FileToIndex? in
-      if sourceFiles.contains(uri) {
-        // If this is a source file, just index it.
-        return .indexableFile(uri)
-      }
-      // Otherwise, see if it is a header file. If so, index a main file that that imports it to update header file's
-      // index.
-      // Deterministically pick a main file. This ensures that we always pick the same main file for a header. This way,
-      // if we request the same header to be indexed twice, we'll pick the same unit file the second time around,
-      // realize that its timestamp is later than the modification date of the header and we don't need to re-index.
-      let mainFile = index.checked(for: .deletedFiles)
-        .mainFilesContainingFile(uri: uri, crossLanguage: false)
-        .sorted(by: { $0.stringValue < $1.stringValue }).first
-      guard let mainFile else {
-        return nil
+    let deletedFilesIndex = index.checked(for: .deletedFiles)
+    let modifiedFilesIndex = index.checked(for: .modifiedFiles)
+    let filesToReIndex =
+      await files
+      .asyncFilter {
+        // First, check if we know that the file is up-to-date, in which case we don't need to hit the index or file
+        // system at all
+        if !indexFilesWithUpToDateUnits, await indexStoreUpToDateTracker.isUpToDate($0) {
+          return false
+        }
+        if case .waitingForPreparation = inProgressIndexTasks[$0] {
+          // We haven't started preparing the file yet. Scheduling a new index operation for it won't produce any
+          // more recent results.
+          return false
+        }
+        return true
+      }.compactMap { (uri) -> FileToIndex? in
+        if sourceFiles.contains(uri) {
+          if !indexFilesWithUpToDateUnits, modifiedFilesIndex.hasUpToDateUnit(for: uri) {
+            return nil
+          }
+          // If this is a source file, just index it.
+          return .indexableFile(uri)
+        }
+        // Otherwise, see if it is a header file. If so, index a main file that that imports it to update header file's
+        // index.
+        // Deterministically pick a main file. This ensures that we always pick the same main file for a header. This way,
+        // if we request the same header to be indexed twice, we'll pick the same unit file the second time around,
+        // realize that its timestamp is later than the modification date of the header and we don't need to re-index.
+        let mainFile =
+          deletedFilesIndex
+          .mainFilesContainingFile(uri: uri, crossLanguage: false)
+          .sorted(by: { $0.stringValue < $1.stringValue }).first
+        guard let mainFile else {
+          return nil
```

Comment:
Could we log in this case (ie. we have a file that isn't known to the project)? Or did you do that and it is that too spammy?

Reply:
Good idea